composer/callbacks/checkpoint_saver.py

```diff
@@ -320,6 +383,43 @@ def __init__(
 
         self.start_batch = None
 
+        self.remote_uploader = None
+        backend, _, _ = parse_uri(save_folder)
+        self.symlink_file_tasks: list[tuple[str, str]] = []
+        self.rank_saves_remote_symlinks: bool = False
+        self.tmp_dir_for_symlink = tempfile.TemporaryDirectory()
+        self.num_concurrent_uploads = num_concurrent_uploads
+        self.symlink_upload_executor = None
+        self.is_remote_upload_failed = None
+        self.symlink_upload_futures = []
+        self.upload_timeout_in_seconds = upload_timeout_in_seconds
+        # Allow unit test to override this to make it faster
+        self._symlink_upload_wait_before_next_try_in_seconds = 30.0
+        self.pid = os.getpid()
+        self.symlink_count = 0
```

Comment:
Does this need to get saved in state_dict for resumption?

Reply:
No, it's just used to distinguish the local symlink file saved in the temporary path.